src/api/java/baritone/api/schematic/SubstituteSchematic.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * This file is part of Baritone.
+ *
+ * Baritone is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Baritone is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Baritone.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package baritone.api.schematic;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.properties.IProperty;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.init.Blocks;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+public class SubstituteSchematic extends AbstractSchematic {
+
+    private final ISchematic schematic;
+    private final Map<Block, List<Block>> substitutions;
+
+    public SubstituteSchematic(ISchematic schematic, Map<Block,List<Block>> substitutions) {
+        super(schematic.widthX(), schematic.heightY(), schematic.lengthZ());
+        this.schematic = schematic;
+        this.substitutions = substitutions;
+    }
+
+    @Override
+    public boolean inSchematic(int x, int y, int z, IBlockState currentState) {
+        return schematic.inSchematic(x, y, z, currentState);
+    }
+
+    @Override
+    public IBlockState desiredState(int x, int y, int z, IBlockState current, List<IBlockState> approxPlaceable) {
+        IBlockState desired = schematic.desiredState(x, y, z, current, approxPlaceable);
+        Block desiredBlock = desired.getBlock();
+        if (!substitutions.containsKey(desiredBlock)) {
+            return desired;
+        }
+        List<Block> substitutes = substitutions.get(desiredBlock);
+        if (substitutes.contains(current.getBlock()) && !current.getBlock().equals(Blocks.AIR)) {// don't preserve air, it's almost always there and almost never wanted
+            return withBlock(desired, current.getBlock());
+        }
+        for (Block substitute : substitutes) {
+            if (substitute.equals(Blocks.AIR)) {
+                return Blocks.AIR.getDefaultState(); // can always "place" air
```

Comment:
I don't know for sure, but perhaps it would fix things if this were instead `return current.getBlock() instanceof BlockAir ? current : Blocks.AIR.getDefaultState();`? Very unsure of this though, feel free to ignore if I'm wrong

Reply:
I'm not sure about this because I
1. Don't want to special-case too many blocks (i.e. when they say "replace it with cave_air" Baritone should try doing so)
2. When used as a "build as long as you can and proceed anyway if you can't" this will likely fail in caves, because people likely didn't put `cave_air` as a substitute.

But since I consider the 2. point to be more important, I'll use what you suggested rather than the alternative for 1. (`return current.getBlock() instanceof BlockAir ? current : substitute;`)